### PR TITLE
refactor SmallBuffer

### PR DESCRIPTION
**Pull request by WalterBright**
1. add `scope` where appropriate
2. replace `T` with `Element` since it is creating an array of elements
3. check for size overflow
4. add missing documentation
5. add `@nogc`

### CODE REVIEW

File: compiler/src/dmd/common/string.d [hunk 1]
```diff
@@ -13,35 +13,43 @@ module dmd.common.string;
...
+    scope this(size_t len, return scope Element[] buffer)
     {
         if (len <= buffer.length)
         {
             _extent = buffer[0 .. len];
         }
         else
         {
-            _extent = (cast(typeof(_extent.ptr)) malloc(len * _extent[0].sizeof))[0 .. len];
+	    assert(len < sizeof.max / Element.sizeof);
```
Comment (dkorpel) [inline compiler/src/dmd/common/string.d:51]:
```suggestion
            assert(len < sizeof.max / Element.sizeof);
```
detab

Comment (WalterBright) [reply compiler/src/dmd/common/string.d]: done

**Review by RazvanN7 — APPROVED** [review]

File: compiler/src/dmd/common/string.d [hunk 2]
```diff
@@ -13,35 +13,43 @@ module dmd.common.string;
...
+    scope this(size_t len, return scope Element[] buffer)
     {
         if (len <= buffer.length)
         {
             _extent = buffer[0 .. len];
         }
         else
         {
-            _extent = (cast(typeof(_extent.ptr)) malloc(len * _extent[0].sizeof))[0 .. len];
+            assert(len < sizeof.max / Element.sizeof);
```
Comment (thewilsonator) [inline compiler/src/dmd/common/string.d:51]: `sizeof.max` is bogus, you meant `size_t`. Not sure how this passed CI, or why it hasn't bee a problem until now.

File: compiler/src/dmd/common/string.d [hunk 3]
```diff
@@ -54,16 +62,22 @@ struct SmallBuffer(T)
...
         if (len <= _extent.length)
         {
-            _extent = _extent[0 .. len];
+            _extent = _extent[0 .. len]; // reuse existing storage
         }
         else
         {
             __dtor();
-            _extent = (cast(typeof(_extent.ptr)) malloc(len * _extent[0].sizeof))[0 .. len];
+            assert(len < sizeof.max / Element.sizeof);
```
Comment (thewilsonator) [inline compiler/src/dmd/common/string.d:79]: ditto

**Pull request merged**